framework/decode/decode_json_util.h

```diff
@@ -0,0 +1,244 @@
+template <typename DecodedType, typename OutputDecodedType = DecodedType>
+void FieldToJson(nlohmann::ordered_json&                               jdata,
+                 const PointerDecoder<DecodedType, OutputDecodedType>& data,
+                 const util::JsonOptions&                              options = util::JsonOptions())
+{
+    FieldToJson(jdata, &data, options);
+}
+
+template <>
+void FieldToJson(nlohmann::ordered_json&                   jdata,
+                 const PointerDecoder<uint64_t, uint64_t>& data,
+                 const util::JsonOptions&                  options);
+
+template <typename DecodedType>
+void FieldToJson(nlohmann::ordered_json&                  jdata,
+                 const StructPointerDecoder<DecodedType>* data,
+                 const util::JsonOptions&                 options = util::JsonOptions())
+{
+    if (data)
+    {
+        const auto meta_struct = data->GetMetaStructPointer();
+        const auto length      = data->GetLength();
+        if (data->IsArray())
+        {
+            for (size_t i = 0; i < length; ++i)
+            {
+                FieldToJson(jdata[i], &meta_struct[i], options);
+            }
+        }
+        else if (length == 1)
+        {
+            FieldToJson(jdata, meta_struct, options);
+        }
+    }
+}
+
+// Similar to above but DecodedType is pointed-to
+template <typename DecodedType>
+void FieldToJson(nlohmann::ordered_json&                   jdata,
+                 const StructPointerDecoder<DecodedType*>* data,
+                 const util::JsonOptions&                  options = util::JsonOptions())
+{
+    if (data)
+    {
+        const auto meta_struct = data->GetMetaStructPointer();
+        const auto length      = data->GetLength();
+        if (data->IsArray())
+        {
+            for (size_t i = 0; i < length; ++i)
+            {
+                FieldToJson(jdata[i], meta_struct[i], options);
+            }
+        }
+        else if (length == 1)
+        {
+            FieldToJson(jdata, *meta_struct, options);
+        }
+    }
+}
+
+template <typename THandle>
+void HandleToJson(nlohmann::ordered_json&              jdata,
+                  const HandlePointerDecoder<THandle>* data,
+                  const util::JsonOptions&             options = util::JsonOptions())
+{
+    if (data && data->GetPointer())
+    {
+        const auto decoded_value = data->GetPointer();
+        const auto length        = data->GetLength();
+
+        if (data->IsArray())
+        {
+            for (size_t i = 0; i < length; ++i)
+            {
+                HandleToJson(jdata[i], decoded_value[i], options);
+            }
+        }
+        else if (length == 1)
+        {
+            HandleToJson(jdata, *decoded_value, options);
+        }
+    }
+}
+
+/// @brief Thunk to HandleToJson for manual conversion functions which forget to
+/// use that for the array form.
+template <typename THandle>
```

Comment:
It is actually cleaner in codegen if all the `XToJson()` functions which are distinguishable by argument type use the plain `FieldToJson` name.